### PR TITLE
Rework halide_benchmark.h to use an adaptive iteration technique

### PR DESCRIPTION
**Pull request by steven-johnson**
This is mimicking a (very) simplified version of the approach used by GoogleTest's benchmark facility: instead of requiring the caller to specify explicit samples-and-iterations, we set a target minimum runtime (0.5 seconds), start with one iteration, scaling up the iteration count based on actual timing observed. (There are also knobs to control the min and max iters for pathological cases).

This tends to make most microbenchmarks simpler and more reliable (no guesswork about decent values), and also makes the results more familiar to users who are familiar with GTest's benchmarking facility (especially when used via RunGen).

Note that the "old" benchmark function is still present (for now); RunGen and test/performance have been converted to "new" style as proof-of-concept. (If this PR is approved then other uses will be converted in subsequent pulls.)

### CODE REVIEW

**Comment by abadams** [comment]
There are two parameters to benchmark() - one runs the test multiple times to account for clock resolution issues, and the other takes the min over multiple trials to account for variance across runs due to additive timing noise from the system swapping out your process, or flipping your thread across sockets, or similar. 

Clock resolution only matters for the fastest tests - mostly we want to address the variance problem.  Iterating up to a minimum time doesn't really improve things in the right way from a statistical perspective. It makes fast tests needlessly slow, and slow tests flaky. Any sort of early-out based on total runtime so far is also very susceptible to the kind of additive noise we get - random stalls - which can bump the runtime of a fast test across the threshold due to a system hiccup, and then you don't take the more iterations that you usually would to discover the error. This was a problem with the linear algebra app benchmarks for a long time.

Adaptive behavior could be good, but you'd want to use something that adapted to the variance seen in the runs so far, rather than the total time. E.g. one trick might be to keep doubling iterations until the test takes at least 5ms, and then run trials until the third best runtime seen is less than 1% slower than the best runtime seen. That's the sort of thing I usually do by hand when benchmarking something manually. @dsharletg might have some thoughts.

IMO GTest sucks for benchmarking.

**Comment by steven-johnson** [comment]
Message received. 

My underlying goal here was to make our benchmarking usage/results more intuitively similar to what people see from GTest results, as that's likely to be what most of our Google-internal users expect (and/or require in some cases). It may be that we need to continue to offer both styles of benchmark (with RunGen offering a choice via flag?) , or it may be that the adaptive style can address your concerns. Probably simpler to brainstorm this F2F.

**Comment by abadams** [comment]
I think there's probably some adaptive strategy that makes sense, we just need to select it carefully. It occurs to me that my proposal works if the timing noise is indeed strictly positive, but if you get a negative blip for some reason it will loop forever.

**Comment by steven-johnson** [comment]
> [random stalls] was a problem with the linear algebra app benchmarks

Was this fixed via tweaking the benchmark args, or the app itself? (I'd like to try reproing the stalls if possible for testing purposes to see if the adaptive approach can fix it)

> negative blip

Yeah, looking at the public GTest benchmark implementation, looks like they have extra logic to ensure that the scale factor never drops below 1.0.

File: tools/halide_benchmark.h [hunk 1]
```diff
@@ -20,19 +39,129 @@ namespace Tools {
...
+    if (times[0] * iters_per_sample * kMinSamples >= min_time) {
+      break;
+    }
+    // Use an estimate based on initial times to converge faster.
+    double next_iters = std::max(min_time / std::max(times[0] * kMinSamples, 1e-9),
+                                 iters_per_sample * 2.0);
+    iters_per_sample = (uint64_t)(next_iters + 0.5);
+  }
+
+  // While we aren't accurate enough (and we have time and iterations remaining)
```
Comment (abadams) [inline tools/halide_benchmark.h]: I don't think this condition is quite right. I think it should keep taking samples until it has the required accuracy, even if that goes beyond min_time. I was imagining:

(times[0] * accuracy < times[kMinSamples - 1] || total_time < min_time) && result.iterations < max_iters && total_time < max_time

Comment (steven-johnson) [reply tools/halide_benchmark.h]: Yeah, good point: except there is no max_time setting. (max_iters is the "hard stop" limit.) Will fix.

**Comment by steven-johnson** [comment]
- Keep running while we have more time, even if accurate enough
- Set an upper time limit to guard against pathology
- Return the measured accuracy and report it in RunGen

**Comment by steven-johnson** [comment]
Monday morning review ping

File: tools/halide_benchmark.h [hunk 2]
```diff
@@ -20,19 +39,140 @@ namespace Tools {
...
+// for real-world use. For now, callers using this to benchmark GPU
+// code should measure with extreme caution.
+
+constexpr uint64_t kBenchmarkMaxIterations = 1000000000;
+
+struct BenchmarkConfig {
+  // Attempt to use this much time (in seconds) for the meaningful samples
+  // taken; initial iterations will be done to find an iterations-per-sample
+  // count that puts the total runtime in this ballpark.
+  double min_time{0.5};
```
Comment (abadams) [inline tools/halide_benchmark.h]: We should see how low we can get this as a reasonable default.

File: tools/halide_benchmark.h [hunk 3]
```diff
@@ -20,19 +39,140 @@ namespace Tools {
...
+
+  // Measured accuracy between the best and third-best result. 
+  // Will be <= config.accuracy unless max_iters is exceeded.
+  double accuracy;
+
+  operator double() const { return wall_time; }
+};
+
+inline BenchmarkResult benchmark(std::function<void()> op, const BenchmarkConfig& config = {}) {
+  BenchmarkResult result{0, 0, 0};
```
Comment (abadams) [inline tools/halide_benchmark.h]: two-space indentation in this file

**Comment by abadams** [comment]
lgtm apart from my two comments above

**Comment by steven-johnson** [comment]
Woops, I forgot to fix the indentation. Will fix directly to master.

**Comment by abadams** [comment]
Bah, the merge was a misclick anyway - I thought I was merging a different branch.

I wanted to check the total runtime of test_performance before merging this. Looks like it's 106 seconds, which is fine.

**Comment by steven-johnson** [comment]
Ah, gotcha. Feel free to back out.

**Comment by abadams** [comment]
Nah, that plus the indentation were the last two things.

**Pull request merged**